src/Workspaces/Remote/ServiceHub/Services/CodeAnalysisService_CodeRefactoring.cs

```diff
@@ -0,0 +1,55 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeRefactorings;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.PasteTracking;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeAnalysis.Remote
+{
+    internal partial class CodeAnalysisService : IRemoteCodeRefactoringService
+    {
+        public Task<bool> HasRefactoringsAsync(PinnedSolutionInfo solutionInfo, DocumentId documentId, TextSpan textSpan, CancellationToken cancellationToken)
+        {
+            return RunServiceAsync(async () =>
+            {
+                using (UserOperationBooster.Boost())
+                {
+                    using var pasteTrackingServiceRegistration = RemotePasteTrackingService.RegisterCallback(new PasteTrackingServiceCallback(EndPoint, cancellationToken));
```

Comment:
so, if HasrefactoringsAsync just passed the paste span, we could literally just register that span locally and avoid the entire need for weakrefs and all that complexith here.

Reply:
Done